25.8 Antalya: Fix joins with Iceberg tables

ianton-ru opened the pull request:
<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix joins with Iceberg tables
Solved #1063 (I hope)

### Documentation entry for user-facing changes
* Joins works with Iceberg tables, not only with table functions
* Fix for local columns in ORDER BY/GROUP BY
* WHERE/PREWHERE condition for subquery
* More tests

### CI/CD Options
#### Exclude tests:
- [ ] <!---ci_exclude_fast--> Fast test
- [ ] <!---ci_exclude_integration--> Integration Tests
- [ ] <!---ci_exclude_stateless--> Stateless tests
- [ ] <!---ci_exclude_stateful--> Stateful tests
- [ ] <!---ci_exclude_performance--> Performance tests
- [x] <!---ci_exclude_asan--> All with ASAN
- [x] <!---ci_exclude_tsan--> All with TSAN
- [x] <!---ci_exclude_msan--> All with MSAN
- [x] <!---ci_exclude_ubsan--> All with UBSAN
- [x] <!---ci_exclude_coverage--> All with Coverage
- [ ] <!---ci_exclude_aarch64|arm--> All with Aarch64
- [x] <!---ci_exclude_regression--> All Regression
- [ ] <!---no_ci_cache--> Disable CI Cache

#### Regression jobs to run:
- [ ] <!---ci_regression_common--> Fast suites (mostly <1h)
- [ ] <!---ci_regression_aggregate_functions--> Aggregate Functions (2h)
- [ ] <!---ci_regression_alter--> Alter (1.5h)
- [ ] <!---ci_regression_benchmark--> Benchmark (30m)
- [ ] <!---ci_regression_clickhouse_keeper--> ClickHouse Keeper (1h)
- [ ] <!---ci_regression_iceberg--> Iceberg (2h)
- [ ] <!---ci_regression_ldap--> LDAP (1h)
- [ ] <!---ci_regression_parquet--> Parquet (1.5h)
- [ ] <!---ci_regression_rbac--> RBAC (1.5h)
- [ ] <!---ci_regression_ssl_server--> SSL Server (1h)
- [ ] <!---ci_regression_s3--> S3 (2h)
- [ ] <!---ci_regression_tiered_storage--> Tiered Storage (2h)

ianton-ru committed:
Join for tables

ianton-ru committed:
Some fixes

ianton-ru committed:
Add test for tables

ianton-ru commented:
Failed tests:
`03572_export_replicated_merge_tree_part_to_object_storage` - looks like intersection in object_storage_cluster and part export
`02995_new_settings_history`:
```
2025-10-15 14:20:45 +PLEASE ADD THE NEW SETTING TO SettingsChangesHistory.cpp: export_merge_tree_part_overwrite_file_if_exists WAS ADDED
2025-10-15 14:20:45 +PLEASE ADD THE NEW SETTING TO SettingsChangesHistory.cpp: allow_experimental_export_merge_tree_part WAS ADDED
2025-10-15 14:20:45 +PLEASE ADD THE NEW SETTING TO SettingsChangesHistory.cpp: allow_experimental_export_merge_tree_part WAS ADDED
```
`01079_parallel_alter_detach_table_zookeeper` - looks like flap

ianton-ru committed:
Fix stateless tests

mkmkme commented on src/Storages/IStorageCluster.cpp:
```diff
@@ -501,25 +528,31 @@ QueryProcessingStage::Enum IStorageCluster::getQueryProcessingStage(
+        if (query_node.hasWhere() || query_node.hasPrewhere())
+        {
+            CollectUsedColumnsForSourceVisitor collector_where(table_function_node, context, true);
+            if (query_node.hasPrewhere())
```
It's quite odd that the original code didn't take prewhere into account...

If those are actually needed, we need to place a mental pin here that after backporting prewhere+row_policy we'll need to add `hasRowLevelFilter` here as well.

mkmkme commented on tests/integration/test_database_iceberg/test.py:
```diff
@@ -609,3 +611,86 @@ def test_table_with_slash(started_cluster):
     assert node.query(f"SELECT * FROM {CATALOG_NAME}.`{root_namespace}.{table_encoded_name}`") == "\\N\tAAPL\t193.24\t193.31\t('bot')\n"
+
+
+def test_cluster_joins(started_cluster):
```
This test seems to be passing on the current Antalya branch:

```
test_database_iceberg/test.py::test_cluster_joins PASSED  
```

Is it expected?

ianton-ru replied:
Nice catch, test doesn't cover case with local right table, thanks!

ianton-ru replied:
And not on cluster...

arthurpassos replied:
Can you add a test for the `CROSS JOIN` case as well?

ianton-ru replied:
Done.

ianton-ru committed:
Fix typo

ianton-ru committed:
Add tests with local table

ianton-ru committed:
ColumnConstable

ianton-ru committed:
Merge branch 'feature/antalya-25.8/column_constable' into feature/antalya-25.8/s3cluster_global_join_fixes

ianton-ru committed:
Fix build with sanitizer

ianton-ru committed:
Forgotten change

ianton-ru committed:
Iceberg cross join

ianton-ru committed:
Fix info about joins

ianton-ru committed:
Merge branch 'antalya-25.8' into feature/antalya-25.8/s3cluster_global_join_fixes

ianton-ru committed:
Turn off contable column

ianton-ru committed:
Revert "ColumnConstable"

This reverts commit bff8bbaf0dfe824cde521e1ed4bfea26f3dbf885.

ianton-ru committed:
Merge branch 'antalya-25.8' into feature/antalya-25.8/s3cluster_global_join_fixes

ianton-ru commented:
Remove unused code

arthurpassos commented:
@ianton-ru can you give me a bit more context on what the issue was in each specific bullet point and how it was fixed?

For example:

> Joins works with Iceberg tables, not only with table functions

Why wasn't it working? What was missing? And what did you do to get it working?

Same for the others.

ianton-ru commented:
> Joins works with Iceberg tables, not only with table functions

When I search tree node to make subquery to replicas, I have searched only `QueryTreeNodeType::TABLE_FUNCTION` before (in https://github.com/Altinity/ClickHouse/pull/972).
Now `QueryTreeNodeType::TABLE` is processed too.

> Fix for local columns in ORDER BY/GROUP BY

`ORDER BY` condition sent to replica nodes and got error about unknown identifiers when it with local columns, now condition works only on local node.

> WHERE/PREWHERE condition for subquery
In first realization I removed WHERE/PREWHERE condition in subquery when is contains local columns.
```
SELECT ice.x, local.y
  FROM iceberg.table AS ice
  JOIN local.table AS local
  ON ice.key=local.key
WHERE
  ice.x >= 10 AND local.y >= 20
```
Befor subquery was like
```
SELECT ice.x
  FROM iceberg.table AS ice
```
Now `removeExpressionsThatDoNotDependOnTableIdentifiers` is calles, and subquery with part of WHERE condition ()
```
SELECT ice.x
  FROM iceberg.table AS ice
WHERE
  ixe.x >= 10
```

Also fixed bugs founded by Alsu:

* `CROSS JOIN` (has separate node type in query tree)
* Lost strings (forgotten `convertToFullColumnIfConst` in `StorageObjectStorageSource::generate`)
* Incorrect work when table function in joined with itself
```
SELECT * FROM iceberg('path') JOIN iceberg('path') ...
```
storage reused in this case, but I tried to save some info about query (`has_join`) inside storage object. As result was attempt to process right table as left table with strange errors.

arthurpassos commented on src/Storages/IStorageCluster.cpp:
```diff
@@ -492,38 +530,58 @@ void ReadFromCluster::initializePipeline(QueryPipelineBuilder & pipeline, const
-        auto table_function_node = table_function_searcher.getNode();
-        if (!table_function_node)
-            throw Exception(ErrorCodes::LOGICAL_ERROR, "Can't find table function node");
+    SearcherVisitor table_function_searcher({QueryTreeNodeType::TABLE, QueryTreeNodeType::TABLE_FUNCTION}, context);
```
Nit: is this still a `table_FUNCTION_visitor`?

ianton-ru replied:
Table and table function. Try to make better name. Actually I try to get left expression of join here, something like `left_table_expression_searcher`.

arthurpassos commented on src/Storages/IStorageCluster.h:
```diff
@@ -105,8 +105,14 @@ class IStorageCluster : public IStorage
+        bool has_local_columns_in_where = false;
+    };
+
+    static QueryTreeInfo getQueryTreeInfo(QueryTreeNodePtr query_tree, ContextPtr context);
```
Please add docs explaining what this method does

arthurpassos commented on src/Storages/IStorageCluster.cpp:
```diff
@@ -492,38 +530,58 @@ void ReadFromCluster::initializePipeline(QueryPipelineBuilder & pipeline, const
+    auto & query_node = query_tree->as<QueryNode &>();
+    if (query_node.hasWhere() || query_node.hasPrewhere())
+    {
         CollectUsedColumnsForSourceVisitor collector_where(table_function_node, context, true);
```
I understand `CollectUsedColumnsForSourceVisitor` is meant to collect all columns for a given "source". As I understand, the source is provided in the constructor. But what makes it local? Doesn't the query_tree object you are passing represent the entire query tree? I don't get it, please educate me

ianton-ru replied:
Yes, `query_tree` contains all, But in different places.
Columns can be in selected list, in where condition, in join condition, etc.
```
SELECT table1.column1,... FROM table1 JOIN table2 ON table1.column2=table2.column2 WHERE table1.column3=...
```
and collector traverses the tree and collect all cases in single list.
Need to all these columns to select list for left table subquery.

arthurpassos replied:
I understand the columns can be in different places, but let's say there are no local columns. Wouldn't `collector_where.getColumns()` be non-empty regardless? Btw, what makes a table local or not local assuming you can't check for its existence on the remote node?

ianton-ru replied:
Actually `CollectUsedColumnsForSourceVisitor` find here all columns from other sources (last boolean flag `collect_columns_from_other_sources_` in constructor).
Columns without source skipped in any case.
So when if where something like `table1.column1 =1 AND table2.column2 = 2 AND (table1.column3 + table2.column4) = 3 AND randUniform(1,6) = 4` result contains only `table2.column2` and `table2.column4`. Actually I don't know is `table2` local or not, but author of query can know. If he used `object_storage_cluster_join_mode='local'`, need to process as if it local.

arthurpassos commented on src/Storages/IStorageCluster.cpp:
```diff
@@ -492,38 +530,58 @@ void ReadFromCluster::initializePipeline(QueryPipelineBuilder & pipeline, const
 
-        // Can't use 'WHERE' on remote node if it contains columns from other sources
+        // SELECT x FROM datalake.table WHERE x IN local.table
+        // Need to modify 'WHERE' on remote node if it contains columns from other sources
```
As I understand, you need to modify it because the cluster nodes might not have those tables. Please add this to your comment

arthurpassos commented on src/Storages/IStorageCluster.cpp:
```diff
@@ -216,49 +223,80 @@ void IStorageCluster::updateQueryWithJoinToSendIfNeeded(
+
+            if (info.has_join || info.has_cross_join)
+            {
+                if (table_function_searcher.getType().value() == QueryTreeNodeType::TABLE_FUNCTION)
```
Do I get it right that you are extracting ONLY the table function from the original AST? Why doesn't it have to be done for tables as well?

arthurpassos replied:
I mean, I understand it is being done for tables when grabbing the left side. But why is it different. Is the query tree different for table function vs regular tables?

ianton-ru replied:
Good point. Looks like lats two blocks work for table function too!
But now need to retest everything.

arthurpassos commented on src/Storages/IStorageCluster.cpp:
```diff
@@ -216,49 +223,80 @@ void IStorageCluster::updateQueryWithJoinToSendIfNeeded(
+                    auto & table_function_ast = table_function->as<ASTFunction &>();
+                    query_tree_distributed->setAlias(table_function_ast.alias);
+                }
+                else if (info.has_join)
```
Do I get it right that the iceberg table is always the left one?

ianton-ru replied:
Yes (actually not only iceberg, s3Cluster too). In other cases code does not come into IStorageCluster.
When join with local table on the left and iceberg on the right clickhouse processes query as local join and only on final stage reads right table without attempts to send left to remote nodes as part of the subquery, so I did not touch that case.

arthurpassos replied:
Hmm... So in a nutshell, there are two cases: iceberg / s3cluster table on the left, in which we need to prevent local tables from reaching remote nodes. and the other case, in which the local is resolved locally before reaching this code path?

ianton-ru replied:
Yes. In this code left table always s3Cluster/icebergCluster or Iceberg table.

arthurpassos commented on src/Storages/IStorageCluster.cpp:
```diff
@@ -216,49 +223,80 @@ void IStorageCluster::updateQueryWithJoinToSendIfNeeded(
+                }
+                else
+                {
+                    SearcherVisitor join_searcher({QueryTreeNodeType::CROSS_JOIN}, context);
```
Shouldn't this become a method in QueryNode? For example: `QueryNode::getCrossJoinTree`

ianton-ru replied:
All others methods in QueryNode don't search on request, but returns pre-filled fields for `children` vector. Now is very specific case, not sure that for this single case need to make and try to fill this position in all queries.
May be it make sense, but need to review all cross-join-related code and make refactoring.

arthurpassos commented on src/Storages/IStorageCluster.cpp:
```diff
@@ -216,49 +223,80 @@ void IStorageCluster::updateQueryWithJoinToSendIfNeeded(
+                    auto cross_join_node = join_searcher.getNode();
+                    if (!cross_join_node)
+                        throw Exception(ErrorCodes::LOGICAL_ERROR, "Can't find CROSS JOIN node");
+                    query_tree_distributed = cross_join_node->as<CrossJoinNode>()->getTableExpressions()[0]->clone();
```
Does [0] mean `left`?

ianton-ru replied:
Yes. CROSS JOIN has a vector of expression, can have more than two sources. So element 0 is a left. Can't be empty, as I understand.

arthurpassos replied:
Perhaps just add a comment, but not required

ianton-ru replied:
Add comment

arthurpassos reviewed (commented):
Given my ignorance on the subject, I did my best for the 1st pass. Mostly questions so I can better understand the code and do a follow up. I beg for patience :)

ianton-ru committed:
Fixes after review

ianton-ru committed:
Add tests

ianton-ru committed:
Merge branch 'antalya-25.8' into feature/antalya-25.8/s3cluster_global_join_fixes

arthurpassos reviewed (approved):
LGTM
